driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java

```diff
@@ -221,7 +231,9 @@ public Connection borrowConnection(long timeout, TimeUnit unit) throws Connectio
         }
 
         int connectionCount = open.get() + scheduledForCreation.get();
-        if (connectionCount < options().getMaxConnectionsPerHost(hostDistance)) {
+        if (connectionCount < options().getCoreConnectionsPerHost(hostDistance)) {
+            maybeSpawnNewConnection();
```

Comment:
Yes, that was the intent. Here we know that we already have a connection to execute the current request, so we don't need to be overly aggressive in restoring the core size. As you noted it will be reached over time.